pkg/migration/migrations/public_schema_migration.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package migrations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/migration"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkeys"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/dbdesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/schemadesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/errors"
+)
+
+func publicSchemaMigration(
+	ctx context.Context, _ clusterversion.ClusterVersion, d migration.TenantDeps, _ *jobs.Job,
+) error {
+	// Iterate through database IDs where the public schema ID is 29.
+	// Ignore database with ID 1.
+	// The assumption is that we're done the migration once this query returns no
+	// databases. Is it possible that a user concurrently creates a database
+	// that has a synthetic public schema while this query executes and
+	// that this query doesn't find that database?
+	query := `SELECT "parentID", "parentSchemaID", name, id, crdb_internal_mvcc_timestamp 
+FROM system.namespace WHERE "parentID" != 1 AND "parentSchemaID" = 0 AND name = 'public' ORDER BY "parentID" ASC`
+	rows, err := d.InternalExecutor.QueryIterator(
+		ctx, "get_databases_with_synthetic_public_schemas", nil /* txn */, query,
+	)
```

Comment:
what about an anti-join and then stream through these instead of buffering? Like:

```sql
  SELECT ns_db.name, ns_db.id, ns_db.crdb_internal_mvcc_timestamp
    FROM system.namespace AS ns_db
         INNER JOIN system.namespace AS ns_sc ON (
                                                                ns_db.id
                                                                = ns_sc."parentID"
                                                             )
   WHERE ns_db.id != 1
     AND ns_db."parentSchemaID" = 0
     AND ns_db."parentID" = 0
     AND ns_sc."parentSchemaID" = 0
     AND ns_sc.name = 'public'
     AND ns_sc.id = 29
ORDER BY ns_db.id ASC;
```

Reply:
what's the timestamp for?